### ISSUE TITLE
File descriptor is not released after zip file uploaded via jetty-client

### ISSUE THREAD

**Issue by duanfeixiang** [eclipse/jetty.project#3512]
Every time you upload a file, there will be one more file descriptor, which will not be released.

**Comment by olamy** [comment]
which version? how do you upload the file? which file is leaked exactly? which os are you using?

**Comment by duanfeixiang** [comment]
> which version? how do you upload the file? which file is leaked exactly? which os are you using?

hello，the version isjetty-9.4.14.v20181114, and the code for uploading the file is as shown above

**Comment by olamy** [comment]
can you provide a sample project? which operating system and java version?

**Comment by duanfeixiang** [comment]
> can you provide a sample project? which operating system and java version?
Hello, the version is jetty-9.4.14.v20181114, The operating system is SUSE Linux 11 , java version is 1.8
You can use the sample code shown below， the client is from HttpClient in jetty

RestfulResponse upload(String uri,
                               String fileField, String filePath,
                               Map<String, String> fieldMap,
                               Map<String, String> headers) {
            if (!uri.startsWith("http")) {
                uri = "https://" + defaultIP + ":" + defaultPort + uri;
            }

            Path path = Paths.get(filePath);
            Path fileName = path.getFileName();
            if (fileName != null) {
                try (MultiPartContentProvider contentProvider = new MultiPartContentProvider()) {
                    contentProvider.addFilePart(
                            fileField,
                            fileName.toString(),
                            new PathContentProvider(path),
                            null
                    );
                    fieldMap.forEach((name, value) -> contentProvider.addFieldPart(
                            name,
                            new StringContentProvider(value),
                            null));

                    Request request = client.POST(uri);
                    if (headers != null) {
                        headers.forEach(request::header);
                    }
                    ContentResponse contentResponse = request
                            .scheme("https")
                            .content(contentProvider)
                            .send();
                    RestfulResponse restfulResponse = new RestfulResponse();
                    restfulResponse.setStatus(contentResponse.getStatus());
                    restfulResponse.setResponseJson(contentResponse.getContentAsString());

                    return restfulResponse;
                } catch (InterruptedException | ExecutionException | IOException | TimeoutException e) {
                    LOGGER.error("Http Exception happened: {}", e.getMessage());
                }
            }

            return null;
        }

**Comment by sbordet** [comment]
@duanfeixiang in your code you are not handling any compression (zip, gzip, etc.).

How do you know that a file descriptor is not released?

Does the file descriptor leak happen on the client or on the server?

Are you closing compressed streams after you've used them?

**Comment by duanfeixiang** [comment]
> Are you closing compressed streams after you've used them?

@sbordet The file I uploaded is in zip format and is provided to jetty using MultiPartContentProvider.All the streams I opened can be determined to have been closed. But I still find that every time I upload a file, the file descriptor will be incremented and it will not be released.

I get the service process PID，then
cd /proc/${PID}/fd
ll | grep "zip"

               Request request = client.POST(uri);
                if (headers != null) {
                    headers.forEach(request::header);
                }
                ContentResponse contentResponse = request
                        .scheme("https")
                        .content(contentProvider)
                        .send();

If the “content(contentProvider)” is commented out in the code, the number of file descriptors will not increase. Once the content is added, it will  increase.

**Comment by sbordet** [comment]
@duanfeixiang will it increase on the client, or on the server?

Are you zipping or gzipping?

What do you use to compress on the client and uncompress on the server?

**Comment by duanfeixiang** [comment]
> @duanfeixiang will it increase on the client, or on the server?
> 
> Are you zipping or gzipping?
> 
> What do you use to compress on the client and uncompress on the server?

Even if I write an url that doesn't exist, the request doesn't actually go to the server, and the file descriptor will still increase. So I define the problem must be on the client side, not related to the server.

I use zipping， not  gzipping

**Comment by sbordet** [comment]
@duanfeixiang there is nothing related to compressing, zipping or gzipping in the client code you showed us.

Your code above is correct and will create a connection to the server that is then pooled.
As such, you will see the number of file descriptors increased by 1 after the request.

If you send other requests, `HttpClient` may open (and pool) other connections.
You can configure how many connections will be pooled and their idle timeout.

Hope the above explains your issue.

If not, we need a complete reproducer (client and server) so that we can try ourselves.

**Comment by duanfeixiang** [comment]
Today, I find the problem. I run the code, and I make breakpoint on the first line of `PathContentProvider.PathIterator.close()`, I find it never been called.
So, I want to know where will the close method be called in the source code？

```java
private class PathIterator:

        public ByteBuffer next() {
          ......
          this.channel = Files.newByteChannel(PathContentProvider.this.filePath, StandardOpenOption.READ);
          ......
        }

        public void close() {
           ......
        }
```

**Comment by sbordet** [comment]
@duanfeixiang, `PathContentProvider.PathIterator.close()` is called from `HttpSender.reset()` when the sender is done with sending the content.

You still have not provided enough information.

Write a simple reproducible case that we can try ourselves and attach it to this issue.

**Comment by duanfeixiang** [comment]
@sbordet 

    MultiPartIterator：
    public ByteBuffer next() {
            while(true) {
                switch(this.state) {
                case FIRST_BOUNDARY:
                    ......
                case HEADERS:
                    MultiPartContentProvider.Part part = (MultiPartContentProvider.Part)MultiPartContentProvider.this.parts.get(this.index);
                    ContentProvider content = part.content;
                    if (content instanceof AsyncContentProvider) {
                        ((AsyncContentProvider)content).setListener(MultiPartContentProvider.this.listener);
                    }

                    this.iterator = content.iterator();
                    this.state = MultiPartContentProvider.State.CONTENT;
                    return part.headers.slice();
                case CONTENT:
                    ......
                case MIDDLE_BOUNDARY:
                    ......
                case LAST_BOUNDARY:
                    ......
                case COMPLETE:
                    ......
                }
            }
        }

Please analyze this line of code. 

“this.iterator = content.iterator();”
Note： content.iterator() always return a new Iterator object.

During the file upload process, the value of the this.iterator is first assigned to PathContentProvider.PathIterator, and then assigned to ByteBufferContentProvider.Iterator, and only ByteBufferContentProvider.Iterator is closed in the close method, so PathContentProvider.PathIterator 
 is not closed, resulting in the reading file handle has never been released.

I think you should be more professional and clear in the process of uploading files using MultiPartContentProvider. Carefully sort out the code execution logic. You should find that the code here is problematic.

**Comment by sbordet** [comment]
@duanfeixiang it's clear now where the problem is, working on a fix. Thanks!

**Comment by sbordet** [comment]
@duanfeixiang can you test the latest fix in #4391 and report if it works for you?

**Comment by duanfeixiang** [comment]
Yes. Now the file descriptor can be released after the file uploading completely. Thanks!
